### PR TITLE
Add resource-linked CSS and JS capabilities.

### PR DESCRIPTION
**Pull request by joelgibson**
This PR proposes a way of adding CSS and Javascript to NeCSuS messages, in a more sustainable and debuggable way that just using inline `<style>` and `<script>` tags.

I'm very happy with the linked stylesheets. The linked Javascript modules also work well, however there is a _little_ amount of magic going on behind them, and I want to make sure this is as clear as possible before proceeding.

***Note:*** The approach to JS has been simplified, see the comments further down.

# What does it look like to use?

The upshot: you can now add a stylesheet and script (as a Javascript module) URL to a NeCSuS message, for instance so that a bot with a custom stylesheet and Javascript behaviour looks like this:

```python
@app.post('/buttonbot')
def buttonbot():
    """ButtonBot is an example of how to use resource-linked CSS and Javascript."""
    return {
        'author': 'ButtonBot',
        'text': '<button class="buttonbot" onclick="ButtonBot.onClick()">Click me!</button>',
        'css': '/static/buttonbot.css',
        'mjs': '/static/buttonbot.mjs',
    }
```

Relative URLs are relative to the bot URL. Absolute URLs may also be used.

The directory structure on the server is

```
server.py
static/
  buttonbot.css
  buttonbot.mjs
```

The CSS could be anything, but here we will target the `.buttonbot` class:

```css
/* buttonbot.css */
.buttonbot { background-color: green !important; }
```

There is a little magic which loads the Javascript module at the exported `NAME`, if an exported `NAME` exists in the module, causing the exports to be exposed as the global `ButtonBot`.

```javascript
// buttonbot.mjs
export const NAME = 'ButtonBot';

export function onClick() {
    alert("Hello from ButtonBot!");
}
```

# What further work needs to be done?

1. Is this the best way to handle modules? I want to reduce collisions but still make it easy to use event handlers, and debug things in the console.
2. Check that there is enough feedback that the students and tutors can debug this not working, and
3. Check that errors further up the page do not affect corrected attempts later on. (I.e. we don't need to force-clear a whole room if there is an error).

### CODE REVIEW

**Comment by jordanhenderson** [comment]
I'd recommend introducing shadow DOM to isolate each loaded module from other elements. 

It may be a good idea to introduce a preload step to verify compilation/syntax errors prior to loading the script into the chat runtime, ie. esbuild and/or lightingcss are good options 😊

**Comment by joelgibson** [comment]
Thanks @jordanhenderson for the feedback!

I considered using Shadow DOM for isolation, but I think this is quite complicated to understand for students who are only just learning Javascript. I considered a Javascript module being attached as a global variable a good in-between point, which keeps things relatively understandable and debuggable in the console, and event handlers can still be easily used which is a nice stepping stone. I also _want_ people to be able to access anything on the page (because some students do some cool whacky stuff with Javascript which affects the whole chat room).

I like the idea of using other tools to check syntax errors and things, so I might experiment around with them for the purpose of providing some feedback in the frontend if there are syntax errors etc.

**Comment by jordanhenderson** [comment]
I would mainly implement shadow DOM as a runtime feature, ie. abstracted away from developers, to ensure there are no conflicts between components/modules which could save some time and isolate CSS from other components.

IE something like a render() or mount() function could create a shadow DOM root isolated to the current chat message or session to ensure the main chat UI isn't impacted by individual components.

ESBuild (ideally invoked on component update) would also be good for individual JS modules to ensure each module is isolated/compiled to an IIFE rather than loading modules into window/global context.

**Comment by joelgibson** [comment]
After chats with @sheabunge and @SpanishPear, I've revised the approach to Javascript to be dead-simple: there is a `js` property, which includes the linked resource in a plain old `<script>` tag on the page. Like the CSS, each Javascript URL will be loaded at most once. One still gets a good debugging experience with proper file references in the console, and global scope means all the students definitions are available in the console without any magic.

The minimal example of both these features becomes something like:

```python
# server.py

@app.post('/buttonbot')
def buttonbot():
    """ButtonBot is an example of how to use resource-linked CSS and Javascript."""
    return {
        'author': 'ButtonBot',
        'text': '<button class="buttonbot" onclick="buttonBotClick()">Click me!</button>',
        'css': '/static/buttonbot.css',
        'js': '/static/buttonbot.js',
    }
```

```css
/* /static/buttonbot.css */
.buttonbot { background-color: green; }
```

```javascript
// /static/buttonbot.js
function buttonBotClick() {
  alert("Hello from ButtonBot!");
}
```

Global scope is dead-simple to understand for the students, and if we want extensions in the future - like a hook which calls a function for each message in the room, and as each message comes into the room - we can use an event-handler-like approach, i.e. in the linked script you would run:

```javascript
Necsus.addHandler('message', function(domElt, msgJson) {...})
```

I have not yet included `Necsus.addHandler` since this can be added afterwards, this PR is just about including CSS and JS (and also making the `css`, `js`, `img`, and `media` keys relative-to-the-bot-url if they are not absolute).

File: client/necsus.js [hunk 1]
```diff
@@ -125,6 +128,33 @@ let app = new Vue({
...
     }
 
+    // Take any new CSS stylesheets which were attached to messages, and hoist them into a <link> in the <head>.
+    for (let {message, domElt} of reachedDom.values()) {
+      if (message.css != null && !ACTIVE_STYLESHEETS.has(message.css)) {
+        let link = document.createElement('link');
+        link.setAttribute('rel', 'stylesheet');
+        link.setAttribute('href', message.css);
+
+        console.log(`Inserting stylesheet into <head> from ${message.id}: ${link.outerHTML}`);
```
Comment (spanishpear) [inline client/necsus.js:138]:
```suggestion
console.log(`Inserting stylesheet '${message.css}' into <head> from ${message.id}: ${link.outerHTML}`);
```

File: client/necsus.js [hunk 2]
```diff
@@ -125,6 +128,33 @@ let app = new Vue({
...
+
+    // Take any new JS scripts which were attached to messages, and hoist them into a <link> in the <head>.
+    for (let {message, domElt} of reachedDom.values()) {
+      if (message.js != null && !ACTIVE_SCRIPTS.has(message.js)) {
+        let script = document.createElement('script')
+        script.setAttribute('src', message.js)
+        script.setAttribute('type', 'text/javascript')
+        script.setAttribute('async', true)
+
+        console.log(`Inserting script into <head> from ${message.id}: ${script.outerHTML}`);
```
Comment (spanishpear) [inline client/necsus.js:152]:
```suggestion
console.log(`Inserting script '${message.js}' into <head> from ${message.id}: ${script.outerHTML}`);
```

File: example_bots/server.py [hunk 3]
```diff
@@ -223,3 +225,17 @@ def pong():
...
+@app.post('/buttonbot')
+def buttonbot():
+    """ButtonBot is an example of how to use resource-linked CSS and Javascript."""
+    return {
+        'author': 'ButtonBot',
+        'text': '''
+            <p>Click this button to run good Javascript: <button class="buttonbot" onclick="buttonBotClick()">Click me!</button></p>
+            <p>Click this button to throw a JS error: <button class="buttonbot-error" onclick="buttonBotError()">Don't click me!</button></p>
+        ''',
+        'css': '/static/buttonbot.css',
```
Comment (spanishpear) [inline example_bots/server.py:239]: should this be `./static/buttonbot.css` ? or is absolute path fine

**Review by spanishpear — COMMENTED** [review]
Looks good! One question on the url format (and some logging additions for my own sanity when debuggigng students later)

File: example_bots/server.py (hunk 3 above)
Comment (joelgibson) [reply example_bots/server.py:239]: The logic is just `urllib.parse.urljoin(bot_url, script_url)`.

```python
>>> from urllib.parse import urljoin

>>> urljoin('https://joels.best.bot/mybot', 'static/script.js')
'https://joels.best.bot/static/script.js'

>>> urljoin('https://joels.best.bot/mybot', '/static/script.js')
'https://joels.best.bot/static/script.js'

>>> urljoin('https://joels.best.bot/mybot', 'https://some.other.site/static/script.js')
'https://some.other.site/static/script.js'
```

**Comment by joelgibson** [comment]
I've removed that logging addition @SpanishPear since the `.outerHTML` I had already has it:

Screenshot 2024-01-07 at 11 33 19 am

I'll merge now

**Pull request merged**